src/widgets/__tests__/index.tests.ts

```diff
@@ -0,0 +1,172 @@
+      }
+      case 'hitsPerPage': {
+        const hitsPerPage = widget as Widgets['hitsPerPage'];
+        return hitsPerPage({
+          container,
+          items: [{ default: true, label: 'def', value: 1 }],
+        });
+      }
+      case 'numericMenu': {
+        const numericMenu = widget as Widgets['numericMenu'];
+        return numericMenu({
+          container,
+          attribute: 'attr',
+          items: [{ label: 'x', start: 1, end: 2 }],
+        });
+      }
+      case 'sortBy': {
+        const sortBy = widget as Widgets['sortBy'];
+        return sortBy({
+          container,
+          items: [{ label: 'x', value: 'x' }],
+        });
+      }
+      case 'analytics': {
+        const analytics = widget as Widgets['analytics'];
+        return analytics({
+          pushFunction() {},
+        });
+      }
+      case 'queryRuleContext': {
+        const queryRuleContext = widget as Widgets['queryRuleContext'];
+        return queryRuleContext({
+          trackedFilters: {
+            facet(values) {
+              return values;
+            },
+          },
+        });
+      }
+      case 'places': {
+        const places = widget as Widgets['places'];
+        return places({
+          container: document.createElement('input'),
+          placesReference: () => ({} as PlacesInstance),
+        });
+      }
+      case 'panel': {
+        const panel = widget as Widgets['panel'];
+        return panel()(widgets.hierarchicalMenu)({
+          container,
+          attributes: ['attr1', 'attr2'],
+        });
+      }
+      default: {
+        return widget({ container, attribute: 'attr' });
+      }
+    }
+  }
+}
+
+describe('widgets', () => {
+  describe('$$type', () => {
+    test('present in every widget', () => {
+      const widgetInstances = initiateAllWidgets();
+
+      widgetInstances.forEach(([name, widget]) =>
+        expect([name, widget.$$type]).toEqual([name, expect.any(String)])
+      );
+    });
+
+    test('starts with ais.', () => {
+      const widgetInstances = initiateAllWidgets();
+
+      widgetInstances.forEach(([name, widget]) =>
+        expect([name, widget.$$type!.substr(0, 4)]).toEqual([name, 'ais.'])
+      );
+    });
+  });
+
+  describe('$$widgetType', () => {
+    test('present in all widgets', () => {
+      const widgetInstances = initiateAllWidgets();
+
+      widgetInstances.forEach(([name, widget]) =>
+        expect([name, widget.$$widgetType]).toEqual([name, expect.any(String)])
+      );
+    });
+
+    test('starts with ais.', () => {
+      const widgetInstances = initiateAllWidgets();
+
+      widgetInstances.forEach(([name, widget]) =>
+        expect([name, widget.$$widgetType!.substr(0, 4)]).toEqual([
+          name,
+          'ais.',
+        ])
+      );
+    });
+
+    test('is the same as name', () => {
+      const widgetInstances = initiateAllWidgets();
+
+      widgetInstances.forEach(([name, widget]) => {
+        if (name === 'panel') {
```

Comment:
These automated tests are hard to read, to maintain and they quickly become conditional. We already cover most of those in each `{widget}-test.ts` files. Why changing the strategy for `$$widgetType`?

Reply:
it's true, but some things should be true for all widgets regardless of what the name is. I found that this gave me more confidence that it is implemented correctly, for example that no typo happened between the test & implementation (which would likely also exist in the test, if it's written manually)

Reply:
Whenever we adds a widget, we should check this, but might forget to do so (which will be caught by tests). I think it's acceptable to do this test for all widgets in this one place in the uniformed way.

Reply:
It's going to be a pain to maintain but as you wish.